File: Stack.js
function Stack(){
	this.dataStore = [];
	this.top = 0;
	this.push = push;
	this.pop = pop;
	this.peek = peek;
	this.toString = toString;
	this.length = length;
	this.empty = empty;
}
function push(element){	
	this.dataStore[this.top++] = element;
}
function pop(){
	if(!this.empty()) return this.dataStore[--this.top];
	return false;
}
function peek(){
	if(!this.empty()) return this.dataStore[this.top-1];
	return false;
}
function toString(){
	var result = "";
	for(var i=0; i<this.dataStore.length; i++){
		result += this.dataStore[i] + "\n";
	}
	return result;
}
function length(){
	return this.top;
}
function empty(){
	if(this.top == 0) return true;
	else return false;
}

module.exports = Stack;

File: Stack.java
class Stack<Item> {

	class Node<Item>{
		Item key;
		Node next;
		Node prev;
		public Node(Item key){
			this.key = key;
		}
	}
	
	Node<Integer> head = new Node<Integer>(0);
	Node<Integer> tail = new Node<Integer>(0);
	int length = 0;
	
	public Stack(){
		this.head.next = this.tail;
		this.tail.prev = this.head;
	}
	
	public void push(Item key){
		Node node = new Node(key);
		node.prev = this.tail.prev;
		this.tail.prev.next = node;
		this.tail.prev = node;
		node.next = this.tail;
		this.length++;
	}
	
	public Node pop(){
		Node node = this.tail.prev;
		this.tail.prev = this.tail.prev.prev;
		this.tail.prev.next = this.tail;
		this.length--;
		return node;
	}
	
	public Node peek(){
		return this.tail.prev;
	}
	
	public boolean empty(){
		return this.head.next == this.tail;
	}

}